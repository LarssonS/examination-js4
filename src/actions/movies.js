import database from '../firebase/firebase';

//ADD_MOVIE

export const addMovie = (movie) => ({
    type: 'ADD_MOVIE',
    movie 
});

export const startAddMovie = (movieData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            title = '',
            description = '',
            director = '',
            genre = '',
            filmWrapper = '',
            raiting = 0,
            createdAt = 0
        } = movieData;
        const movie = { title, description, director, genre, filmWrapper, raiting, createdAt, uid };

        return database.ref(`users/movies`).push(movie).then(() => {
            dispatch(addMovie({ ...movie }));
          });
    };
};

export const removeMovie = ({ id } = {}) => ({
    type: 'REMOVE_MOVIE',
    id
});

export const startRemoveMovie = ({ id } = {}) => {
    return(dispatch) => {
        return database.ref(`users/movies/${id}`).remove().then(() => {
            dispatch(removeMovie({ id }));
        });
    }
};

export const editMovie = ({id, updates}) => ({
    type: 'EDIT_MOVIE',
    id,
    updates
});

export const startEditMovie = ({id, updates}) => {
    return(dispatch) => {
        return database.ref(`users/movies/${id}`).update(updates).then(() => {
           dispatch(editMovie({id, updates})); 
        });
    };
};

export function fetchListenerMovies(ref, callback) {
    ref.on("child_added", (snapshot) => {
            const data = { ...snapshot.val(), id: snapshot.key };
            callback(data);
        });
};

export function removeListenerMovie(ref, callback) {
    ref.on("child_removed", (snapshot) => {
            callback(snapshot.key);
        });
};

export function changedListenerMovie(ref, callback) {
    ref.on("child_changed", (snapshot) => {
            const data = { updates: snapshot.val(), id: snapshot.key };
            callback(data);
        });
};
