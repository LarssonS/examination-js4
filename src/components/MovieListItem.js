import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
//import numeral from 'numeral';

const MovieListItem = ({ id, title, description, director, genre, filmWrapper, raiting, createdAt }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{title}</h3>
        </Link>
        <p>
            {description}
            -
            {moment(createdAt).format('MMMM Do, YYYY')}
        </p>
    </div>
);

export default MovieListItem;