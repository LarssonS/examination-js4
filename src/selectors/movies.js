import moment from 'moment';

//Get visible movies
export default (movies, { text, sortBy, startDate, endDate }) => {
    return movies.filter((movie) => {
        const createdAtMoment = moment(movie.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const titleMatch = movie.title.toLowerCase().includes(text.toLowerCase());
        const genreMatch = movie.genre.toLowerCase().includes(sortBy.toLowerCase());

        return startDateMatch && endDateMatch && titleMatch && genreMatch;
    });
}