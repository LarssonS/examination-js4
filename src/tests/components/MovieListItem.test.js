import React from 'react';
import { shallow } from 'enzyme';
import movies from '../fixtures/movies';
import MovieListItem from '../../components/MovieListItem';

test('should render MovieListItem correctly', () => {
  const wrapper = shallow(<MovieListItem {...movies[0]} />);
  expect(wrapper).toMatchSnapshot();
});
