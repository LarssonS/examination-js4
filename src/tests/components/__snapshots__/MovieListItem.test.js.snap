// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MovieListItem correctly 1`] = `
<div>
  <Card
    body={true}
    tag="div"
  >
    <CardImg
      alt="img"
      src="img"
      tag="img"
      top={true}
      width="100%"
    />
    <CardBody
      tag="div"
    >
      <CardTitle
        tag="h2"
      >
        batman
      </CardTitle>
      <CardSubtitle
        tag="h3"
      >
        action
      </CardSubtitle>
      <CardText
        tag="p"
      >
        superhero
      </CardText>
      <Link
        replace={false}
        to="/movie/1"
      >
        <Button
          color="secondary"
          size="lg"
          tag="button"
        >
          Read More
        </Button>
      </Link>
    </CardBody>
  </Card>
</div>
`;
